scripts/build_images/build.py

```diff
@@ -0,0 +1,205 @@
+            pass
+
+        args = [
+            "docker",
+            "buildx",
+            "build",
+            "--platform",
+            platform,
+            "--no-cache",
+            "--tag",
+            f"localhost:5000/{image}",
+            "--progress",
+            "plain",
+            "--metadata-file",
+            metadata_file.name,
+            # "--load",
+            "--output=type=image,push-by-digest=true,name-canonical=true,push=true",
+        ]
+
+        if base_image and base_image_built:
+            args.extend([
+                # We have some images that rely on other images. They're configured to
+                # use this arg to determine which base image tag to use.
+                "--build-arg",
+                f"BASE_IMAGE_TAG={tag}",
+                "--build-arg",
+                "BASE_IMAGE_REGISTRY=localhost:5000",
+            ])
+
+        if should_push:
+            args.extend([
+                # Only tag it with the registry name if we're going to push.
+                "--tag",
+                image,
+            ])
+
+        args.extend([image_path])
+
+        print(f"Building image {image} for platform {platform}")
+        print_and_run_command(args)
+
+        with open(metadata_file.name) as f:
+            metadata = f.read()
+
+        print(f"Metadata: {metadata.strip()}")
+        digest = json.loads(metadata)["containerimage.digest"]
+
+        if is_base_image:
+            # `buildx` cannot use locally-built images when the
+            # `docker-container` driver is used, and that driver must be used in
+            # order to support `push-by-digest`. To make the base image available,
+            # we'll push it to a local registry.
+            #
+            # https://github.com/moby/buildkit/issues/2343
+            print_and_run_command([
+                "docker",
+                "pull",
+                f"localhost:5000/{image}@{digest}",
+            ])
+            print_and_run_command([
+                "docker",
+                "tag",
+                f"localhost:5000/{image}@{digest}",
+                f"localhost:5000/{image}:{tag}",
+            ])
+            print_and_run_command([
+                "docker",
+                "push",
+                f"localhost:5000/{image}:{tag}",
+            ])
+
+        # Write metadata to the metadata directory
+        if metadata_dir:
+            metadata = json.loads(metadata)
+            build_ref = metadata["buildx.build.ref"]
+
+            # If pushing is enabled, the image name will be a comma-separated list of image names.
+            # We'll replace it with just the plain image name.
+            metadata["image.name"] = image
+
+            # We need a unique name for the metadata file. We'll use the part of the
+            # image name after the last slash, and a hash of the build ref.
+            name_without_scope = image.split("/")[-1]
```

Comment:
You could have a variable for the image prefix `prairielearn/workspace-`. Then we avoid relying on splits like this.

Reply:
I suppose this depends on how easy you want it to run locally. I lean towards as little state / assumption made in this file as possible.

Reply:
That's incompatible with the group of `prairielearn/plbase`, `prairielearn/prairielearn`, `prairielearn/executor`.

I don't think it's a bad thing for our build scripts to be tailored to what we're building. I'm explicitly not trying to make a super-generic thing here.